drv/stm32h7-eth/src/ring.rs

```diff
@@ -0,0 +1,429 @@
+    /// all times.
+    next: Cell<usize>,
+}
+
+impl TxRing {
+    /// Creates a new TX DMA ring out of `storage` and `buffers`.
+    ///
+    /// Note that, because `&'static mut` is not `Copy` and cannot be a reborrow
+    /// (because `'static` is the longest lifetime), you lose access to both
+    /// slices upon calling this. There is no (safe) way to get the pieces back
+    /// out. This is deliberate.
+    ///
+    /// # Panics
+    ///
+    /// If `storage` and `buffers` are not the same length.
+    pub fn new(
+        storage: &'static mut [TxDesc],
+        buffers: &'static mut [Buffer],
+    ) -> Self {
+        assert_eq!(storage.len(), buffers.len());
+        // Drop mutability. We needed the caller to prove exclusive ownership,
+        // but we don't actually need &mut. We assume that both areas of memory
+        // are shared with the DMA controller from this point forward.
+        let (storage, buffers) = (&*storage, &*buffers);
+        // Initialize all TxDesc records to a known state, and in particular,
+        // ensure that they're owned by us (not the hardware).
+        for desc in storage {
+            desc.tdes[3].store(0, Ordering::Release);
+        }
+        Self {
+            storage,
+            buffers,
+            next: Cell::new(0),
+        }
+    }
+
+    /// Returns the base pointer of the `TxDesc` ring. This needs to be loaded
+    /// into the DMA controller so it knows where to look for descriptors.
+    pub fn base_ptr(&self) -> *const TxDesc {
+        self.storage.as_ptr()
+    }
+
+    /// Returns a pointer to the byte just past the end of the `TxDesc` ring.
+    /// This too gets loaded into the DMA controller, so that it knows what
+    /// section of the ring is initialized and can be read. (The answer is "all
+    /// of it.")
+    pub fn tail_ptr(&self) -> *const TxDesc {
+        self.storage.as_ptr_range().end
+    }
+
+    /// Returns the count of entries in the descriptor ring / buffers in the
+    /// pool.
+    pub fn len(&self) -> usize {
+        self.storage.len()
+    }
+
+    pub fn is_next_free(&self) -> bool {
+        let d = &self.storage[self.next.get()];
+        // Check whether the hardware has released this.
+        let tdes3 = d.tdes[3].load(Ordering::Acquire);
+        let own = tdes3 & (1 << TDES3_OWN_BIT) != 0;
+        !own
+    }
+
+    /// Attempts to grab the next unused TX buffer in the ring and deposit a
+    /// packet into it.
+    ///
+    /// If the next buffer in the ring is not holding a pending packet, this
+    /// will borrow it and call `body` with its address. `body` returns the
+    /// number of bytes in the packet; this routine will then set up a pending
+    /// transmit descriptor for that prefix of the buffer, and return
+    /// `Some(len)`.
+    ///
+    /// If the next buffer in the ring is pending, that means we have run out of
+    /// TX ring slots. In that case, this will return `None` without invoking
```

Comment:
The docstring seems out of date, since this returns a `Result`, not an `Option`.

Reply:
Fixing!